feat(cli): allow for toml files in --config arg, closes #7859

Pietagorh opened the pull request:
<!--
Before submitting a PR, please read https://github.com/tauri-apps/tauri/blob/dev/.github/CONTRIBUTING.md#pull-request-guidelines

1. Give the PR a descriptive title.

  Examples of good title:
    - fix(windows): fix race condition in event loop
    - docs: update example for `App::show`
    - feat: add `Window::set_fullscreen`

  Examples of bad title:
    - fix #7123
    - update docs
    - fix bugs

2. If there is a related issue, reference it in the PR text, e.g. closes #123.
3. If this change requires a new version, then add a change file in `.changes` directory with the appropriate bump, see https://github.com/tauri-apps/tauri/blob/dev/.changes/README.md
4. Ensure that all your commits are signed https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits
5. Ensure `cargo test` and `cargo clippy` passes.
6. Propose your changes as a draft PR if your work is still in progress.
-->
closes #7859 
## Why
1. can already use toml for the base config, just bringing parity to the --config feat
2. pretty small change

## Tests
- [x] `cargo test`
- [x] `cargo clippy`

amrbashir commented on crates/tauri-cli/src/lib.rs:
```diff
@@ -63,10 +63,17 @@ impl FromStr for ConfigValue {
-        )?))
+        let raw = &read_to_string(&path)
+            .with_context(|| format!("invalid configuration at file {config}"))?;
+        if let Some(extension) = path.extension() {
+          match extension.to_str().expect("invalid unicode in file extension") {
+            "toml" => Ok(Self(::toml::from_str(raw)?)),
+            "json" | "json5" => Ok(Self(serde_json::from_str(raw)?)),
+            _ => anyhow::bail!("unsupported file type")
+          }
+        } else {
+          anyhow::bail!("invalid file path (requires file extension)")
+        }
```
we can always default to JSON, and for Json5, we need to use `json5` crate
```suggestion
        match path.extension() {
          Some(ext) if ext == "toml" => Ok(Self(::toml::from_str(raw)?)),
          Some(ext) if ext == "json5" => Ok(Self(json5::from_str(raw)?)),
          // default to JSON
          _ => Ok(Self(serde_json::from_str(raw)?))
        }
```

Pietagorh replied:
right!
isn't the usual behavior to try parsing .json's with json5 as well?

amrbashir replied:
if a json5 file has comments in it, serde_json will not be able to parse it.

amrbashir replied:
I might have misunderstood your question, in that case yeah we could just use `json5` for `.json` files as well.

amrbashir replied:
you also need to update the documentation option to mention JSON5 support as well.

Pietagorh replied:
> I might have misunderstood your question, in that case yeah we could just use `json5` for `.json` files as well.

From what i've seen in tauri-utils, we parse .json with serde, if we get an error we try with json5 and if json5 gives an error we return the serde error
I'll check again later tho

amrbashir replied:
yeah, I guess we can do that as well

amrbashir commented on .changes/toml-config-patch-in-cli.md:
```diff
@@ -0,0 +1,5 @@
+tauri-cli: minor:enhance
+---
+
+File paths to TOML files can now be used in the cli --config arg
```
```suggestion
Add support for passing TOML and JSON5 config files to `--config` arg
```

amrbashir commented on .changes/toml-config-patch-in-cli.md:
```diff
@@ -0,0 +1,5 @@
+---
+tauri-cli: minor:enhance
```
```suggestion
tauri-cli: minor:enhance
"@tauri-apps/cli": minor:enhance
```

Pietagorh commented:
Updated:
- changed what's been requested
- added the feature flag stuff (`config-toml` and `config-json5`)
**i couldn't confirm that tauri-macros actually correctly dispatches the config features cause i don't know how to build the full package** (passing the individual features works tho)

amrbashir commented on crates/tauri-cli/Cargo.toml:
```diff
@@ -149,3 +150,5 @@ native-tls = [
 ]
 native-tls-vendored = ["native-tls", "tauri-bundler/native-tls-vendored"]
 rustls = ["tauri-bundler/rustls", "cargo-mobile2/rustls", "ureq/rustls"]
+config-json5 = ["json5"]
+config-toml = []
```
no need to add these features. This is the CLI and not shipped with the final user app so we can always include them.

Pietagorh replied:
oh ok thanks

amrbashir commented on crates/tauri-macros/Cargo.toml:
```diff
@@ -22,11 +22,12 @@ syn = { version = "2", features = ["full"] }
 tauri-utils = { version = "2.3.0", path = "../tauri-utils" }
+tauri-cli = { version = "2.4.0", path = "../tauri-cli" }
 
 [features]
 custom-protocol = []
 compression = ["tauri-codegen/compression"]
 isolation = ["tauri-codegen/isolation"]
-config-json5 = ["tauri-codegen/config-json5", "tauri-utils/config-json5"]
-config-toml = ["tauri-codegen/config-toml", "tauri-utils/config-toml"]
+config-json5 = ["tauri-codegen/config-json5", "tauri-cli/config-json5", "tauri-utils/config-json5"]
+config-toml = ["tauri-codegen/config-toml", "tauri-cli/config-toml", "tauri-utils/config-toml"]
```
these need to be reverted, tauri-macros doesn't interact with tauri-cli at all.

tauri-cli is just a binary on its own.

Pietagorh committed:
feat(cli): allow for toml and json5 files in --config arg

amrbashir commented:
Thank you

Pietagorh commented:
> Thank you

No, thank you lol you did most of it (thx for showing me how to do it cleanly as well x))
